### PR TITLE
Improvement to Gaussian Splatting workflow

### PR DESCRIPTION
**Pull request by jb-ye**
Hi @kerrj 

As @f-dy discussed with you earlier, this PR is an improvement to the gaussian splatting workflow implemented in nerfstudio. It has two major changes as detailed below. Please feel free to comment. 

## Enhancement to the gaussian splatting training. 

This PR has more aggressive gaussian growth strategy. Based on our internal evaluation, this PR makes the nerfstudio-gs implementation outperforms the official Inria implementation by a little bit on average. In general, we observe the current PR improves more on indoor datasets.

I also experimented three public datasets from mipnerf360 dataset (bicycle, garden, kitchen) and reported the results below:

Nerfstudio-gs Before / After applying this PR
  | psnr | ssim | lpips
-- | -- | -- | --
bicycle | 25.84/26.15 | 0.771/0.790 | 0.179/0.147
garden | 28.38/28.63 | 0.875/0.887 | 0.090/0.070
kitchen | 31.35/32.24 | 0.929/0.941 | 0.096/0.076

For reference, I also cited results from mipnerf360 / gaussian splatting papers:

  | psnr | ssim | lpips
-- | -- | -- | --
bicycle | 24.37/25.25 | 0.685/0.771 | 0.301/0.205
garden | 26.98/27.41 | 0.813/0.868 | 0.170/0.103
kitchen | 32.23/30.317 | 0.920/0.922 | 0.127/0.129

## PLY exporter 

I made changes to allow the PLY exporter produce PLY formatted data compatible with the official Inria version. After export the PLY file from saved checkpoint, one can use the exported file in a web viewer that was originally designed for Inria code.

(_DELETED IN LATEST REVISION_:
As part of the implementation, I also implemented an independent PLY based image renderer without reading nerfstudio checkpoint. This is an effort to simplify potential workflow that are more separate from nerfstudio interface.)

### CODE REVIEW

File: nerfstudio/models/gaussian_splatting.py [hunk 1]
```diff
@@ -363,47 +389,55 @@ def refinement_after(self, optimizers: Optimizers, step):
...
                         self.dup_in_optim(optimizers.optimizers[group], dup_idcs, param, 1)
 
-                # Offset all the opacity reset logic by refine_every so that we don't
-                # save checkpoints right when the opacity is reset (saves every 2k)
-                if self.step % reset_interval > self.num_train_data + self.config.refine_every:
-                    # then cull
-                    deleted_mask = self.cull_gaussians()
+                    # After a guassian is split into two new gaussians, the original one should be also pruned.
+                    splits_mask = torch.cat(
+                        (splits, torch.zeros(nsamps * splits.sum() + dups.sum(), device="cuda", dtype=torch.bool))
```
Comment (pwais) [inline nerfstudio/models/gaussian_splatting.py:405]:
```suggestion
                        (splits, torch.zeros(nsamps * splits.sum() + dups.sum(), device=self.device, dtype=torch.bool))
```

File: nerfstudio/models/gaussian_splatting.py [hunk 2]
```diff
@@ -203,19 +209,26 @@ def populate_modules(self):
         self.step = 0
 
         self.crop_box: Optional[OrientedBox] = None
-        self.back_color = torch.zeros(3)
+        self.back_color = torch.zeros(3, device="cuda")
```
Comment (pwais) [inline nerfstudio/models/gaussian_splatting.py:212]:
```suggestion
        self.back_color = torch.zeros(3, device=self.device)
```

File: nerfstudio/scripts/exporter.py [hunk 3]
```diff
@@ -500,30 +500,48 @@ def main(self) -> None:
...
+
+        # post optimization, it is possible have NaN/Inf values in some attributes
+        # to ensure the exported ply file has finite values, we enforce finite filters.
+        select = np.ones(n, dtype=bool)
+        for k, t in map_to_tensors.items():
+            n_before = np.sum(select)
+            select = np.logical_and(select, np.isfinite(t).all(axis=1))
+            n_after = np.sum(select)
+            if n_after < n_before:
+                print(f"{n_before - n_after} NaN/Inf elements in {k}")
```
Comment (pwais) [inline nerfstudio/scripts/exporter.py:539]:
```suggestion
                CONSOLE.print(f"{n_before - n_after} NaN/Inf elements in {k}")
```

File: nerfstudio/scripts/exporter.py [hunk 4]
```diff
@@ -500,30 +500,48 @@ def main(self) -> None:
...
+
+        if np.sum(select) < n:
+            print(f"values have NaN/Inf in map_to_tensors, only export {np.sum(select)}/{n}")
```
Comment (pwais) [inline nerfstudio/scripts/exporter.py:542]:
```suggestion
            CONSOLE.print(f"values have NaN/Inf in map_to_tensors, only export {np.sum(select)}/{n}")
```

**Review by pwais — COMMENTED** [review]
this is interesting!  FWIW if you plan to continue tweaking the algo, you might consider using the nerfstudio template: https://github.com/nerfstudio-project/nerfstudio-method-template   Not sure how much maintainers expect the ns gsplat impl to diverge from reference / original impl, but at least for researchers isolating a fork of an algo using the template helps make it easier to determine provenance.

**Comment by jb-ye** [comment]
> this is interesting! FWIW if you plan to continue tweaking the algo, you might consider using the nerfstudio template: https://github.com/nerfstudio-project/nerfstudio-method-template Not sure how much maintainers expect the ns gsplat impl to diverge from reference / original impl, but at least for researchers isolating a fork of an algo using the template helps make it easier to determine provenance.

This PR actually makes the implementation a lot more closer to the original implementation. Yes, we plan to further improve the algorithm to address some of its foundational limitations. But I don't see these change will go into the same gaussian-splatting model implemented in nerfstudio.

File: pyproject.toml [hunk 5]
```diff
@@ -63,6 +63,7 @@ dependencies = [
     "trimesh>=3.20.2",
     "timm==0.6.7",
     "gsplat==0.1.0",
+    "meshio==5.3.4",
```
[the remark this answers is not in the document]
Comment (jb-ye) [reply pyproject.toml:66]: I was trying to use open3d to read ply but found that it didn't have interface to read custom attributes. If you know how to do so, I'd be happy to incorporate the change.

File: pyproject.toml [hunk 6]
```diff
@@ -128,6 +129,7 @@ ns-render = "nerfstudio.scripts.render:entrypoint"
 ns-export = "nerfstudio.scripts.exporter:entrypoint"
 ns-dev-test = "nerfstudio.scripts.github.run_actions:entrypoint"
 ns-dev-sync-viser-message-defs = "nerfstudio.scripts.viewer.sync_viser_message_defs:entrypoint"
+ns-gsplat-render = "nerfstudio.scripts.gsplat_render:entrypoint"
```
[the remark this answers is not in the document]
Comment (jb-ye) [reply pyproject.toml:132]: Sounds good to me. But I am not sure how to offer this as a plugin, I would rather exclude this script from the current PR and will do a separate PR for the plugin implementation.

**Comment by jb-ye** [comment]
@pwais I had removed the ply based renderer from the PR to make it simpler. I agree that managing a separate renderer would be a bit difficult.

**Comment by ichsan2895** [comment]
Left = Gaussian Splatting
-  30k steps
-  training time = 36 minutes
-  render time =  18 seconds
-  PSNR at 29k = 24,54
-  SSIM at 29k = 0,8483

0000_GSplat_M60-TAT


Right = Gaussian Splatting DEV
(https://github.com/jb-ye/nerfstudio commits=73241be )
-  30k steps
-  training time = 50 minutes
-  render time = 28 seconds
-  PSNR at 29k = 27,23
-  SSIM at 29k = 0,8873

0000_GSplat-DEV_M60-TAT


* I dont know why WanDB does not record PSNR & SSIM at 30k. It is stop at 29k for `Eval Images Metrics Dict (all images)/psnr` & `Eval Images Metrics Dict (all images)/ssim`

File: pyproject.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (pwais) [reply pyproject.toml]: maybe `trimesh` can? https://github.com/mikedh/trimesh/blob/1d0f548e4a082ee229b62f70aa332a57645176cb/trimesh/exchange/ply.py#L900

**Comment by ichsan2895** [comment]
I also try default gsplat exporter in this PR repo:
`ns-export gaussian-splat --load-config outputs/unnamed/gaussian-splatting/2023-12-30_172430/config.yml --output-dir renders/Train_TankAndTemples/`

image

Soo many floaters, but the color seems right, not washed out & don't have pale color. But it really has many floaters that must be cleaned. Meshlab can't do it.

Where is the train? we must find it.
image

**Comment by ichsan2895** [comment]
Dataset = https://www.3dflow.net/3df-zephyr-reconstruction-showcase/

Left = Gaussian Splatting
* 30k steps
* training time = 68 minutes
* render time =  31 seconds
* PSNR at 29k = 23,63
* SSIM at 29k = 0,8331

0000_GSplat_3DZephyr
0003_GSplat_3DZephyr


Right = Gaussian Splatting DEV
(https://github.com/jb-ye/nerfstudio commits=73241be )
* 30k steps
* training time = 94 minutes
* render time =  34 seconds
* PSNR at 29k = 23,52
* SSIM at 29k = 0,8357

0000_GSplat-DEV_3DZephyr
0003_GSplat-DEV_3DZephyr

I dont know why WanDB does not record PSNR & SSIM at 30k. It is stop at 29k for `Eval Images Metrics Dict (all images)/psnr` & `Eval Images Metrics Dict (all images)/ssim`

**Comment by ichsan2895** [comment]
Gaussian Splatting
* 30k steps
* training time = 20 minutes
* render time =  19 seconds
* PSNR at 29k = 23,27
* SSIM at 29k = 0,841

0003_GSplat
0009_GSplat

Gaussian Splatting DEV
(https://github.com/jb-ye/nerfstudio commits=73241be )
* 30k steps
* training time = 33 minutes
* render time =  24 seconds
* PSNR at 29k = 24,61
* SSIM at 29k = 0,872

0003_GSplat-DEV
0009_GSplat-DEV

**Review by ichsan2895 — COMMENTED** [review]
LGTM

**Comment by ichsan2895** [comment]
From my experiment, Gaussian Splatting-DEV (this implementation) use lower GPU memory & higher GPU utilization than nerfstudio's gsplat in main repo.

#################
image
Dataset = Truck TanksAndTemples
green = nerfstudio's gsplat using 38 GB of VRAM but shorter training duration
cream = this gsplat implementation using only 7 GB of VRAM and utilizer more GPU utilization.

#################
image
Dataset = M60 TanksAndTemples
red = nerfstudio's gsplat using 43 GB of VRAM but shorter training duration
green = this gsplat implementation using only 8,5 GB of VRAM and utilizer more GPU utilization.

**Comment by jb-ye** [comment]
@ichsan2895 Thanks a lot for the benchmark. Just to note that the high memory usage seen in the original implementation is due to the memory fragmentation of gaussian culling. By actively emptying cache of torch, you would save a lot memory. In my PR, I had this change to avoid memory fragmentation.

**Comment by ichsan2895** [comment]
> @ichsan2895 Thanks a lot for the benchmark. Just to note that the high memory usage seen in the original implementation is due to the memory fragmentation of gaussian culling. By actively emptying cache of torch cuda, you would save a lot memory. In my PR, I had this change to avoid memory fragmentation.

So its more better result and also more efficient. What a great improvement. Hopefully merged to main Repo.

**Comment by ichsan2895** [comment]
Thats being said, Color fix is fixed in this commit: `https://github.com/nerfstudio-project@4a54763323041941f43140736d1b76c80a75af7f` (20 december 2023)

But I tested this commit: `https://github.com/nerfstudio-project/a8e6f8fa3fd6c0ad2f3e681dcf1519e74ad2230f` (28 december 2023) and the color of PLY file is still wrong. Greyed out and pale when I open it with antimatter15 WebGL GS viewer

Fortunatelly, the implementation of GSplat fixes it.

**Comment by hardikdava** [comment]
@jb-ye great improvement. I have tried  and really great improvement. Is it possible to break down methods for densification, splitting, opacity reset just like inria 3d gaussian splatting implementation? Then further improvement will be much easier I believe.

**Comment by jb-ye** [comment]
@ichsan2895 I reposted your side-by-side comparison videos on nerfstudio discord [here](https://github.com/jb-ye/nerfstudio/issues/1). I have a paid account where I can upload video up to 100MB.

**Comment by kerrj** [comment]
Hi! Thanks for the changes they look great :). 

- One thing I'm concerned about is the speed difference, having more gaussians makes training ~2x slower. I think we can get away with a bit stricter culling parameters with minimal quality changes, for example here's plots of gaussians and psnr over time on your branch (pink) versus using a 0.1 threshold and culling until 30k (green) on the bicycle dataset. Final PSNR is only .3 different, and it trains about 2x faster. I think it might be good to default to faster parameters like these, and tweaking them is easy for users that want the extra bit of quality.
Screenshot 2024-01-05 at 3 07 57 PM
Screenshot 2024-01-05 at 3 08 08 PM


- As a side note, are you also seeing these PSNR drops every 5k steps while training? I haven't been able to figure out where it's coming from, since it continues even after opacity resetting stops (after 15k steps)

File: nerfstudio/models/gaussian_splatting.py [hunk 7]
```diff
@@ -686,26 +743,8 @@ def get_loss_dict(self, outputs, batch, metrics_dict=None) -> Dict[str, torch.Te
             gt_img = batch["image"]
         Ll1 = torch.abs(gt_img - outputs["rgb"]).mean()
         simloss = 1 - self.ssim(gt_img.permute(2, 0, 1)[None, ...], outputs["rgb"].permute(2, 0, 1)[None, ...])
-        if self.step % 10 == 0:
-            # Before, we made split sh and colors onto different optimizer, with shs having a low learning rate
-            # This is slow, instead we apply a regularization every few steps
-            sh_reg = self.colors_all[:, 1:, :].norm(dim=1).mean()
-            scale_exp = torch.exp(self.scales)
-            scale_reg = (
```
Comment (kerrj) [inline nerfstudio/models/gaussian_splatting.py:694]: This PR removes the scale regularization (taken from PhysGauss paper), I found it's useful for reducing the number of huge spikey gaussians, did you find it impacts quality? It's ok if we need to remove it/make it optional, but if so the parameter `max_gauss_ratio` should be refactored/removed.

**Comment by ichsan2895** [comment]
> Hi! Thanks for the changes they look great :).
> 
>     * One thing I'm concerned about is the speed difference, having more gaussians makes training ~2x slower. I think we can get away with a bit stricter culling parameters with minimal quality changes, for example here's plots of gaussians and psnr over time on your branch (pink) versus using a 0.1 threshold and culling until 30k (green) on the bicycle dataset. Final PSNR is only .3 different, and it trains about 2x faster. I think it might be good to default to faster parameters like these, and tweaking them is easy for users that want the extra bit of quality.
> 
> 
> Screenshot 2024-01-05 at 3 07 57 PM Screenshot 2024-01-05 at 3 08 08 PM
> 
>     * As a side note, are you also seeing these PSNR drops every 5k steps while training? I haven't been able to figure out where it's coming from, since it continues even after opacity resetting stops (after 15k steps)

From my experience:
I sometimes do not occur spikes in `eval all images PSNR`. If it happens, it also happen in gsplat at main nerfstudio repo..

Desolation dataset. Purple is this implementation, blue is nerfstudio's gsplat.
image
3D Zephyr human dataset. Purple is this implementation, magenta is nerfstudio's gsplat. The spikes also happens in both repo.
image

**Comment by jb-ye** [comment]
@kerrj Thanks for feedback. I am okay to set a strict culling parameter to speed up the training by default. 

> * As a side note, are you also seeing these PSNR drops every 5k steps while training? I haven't been able to figure out where it's coming from, since it continues even after opacity resetting stops (after 15k steps)

This is new to me. Let me spend time to investigate a bit. It may be something we didn't expect.

**Comment by jb-ye** [comment]
> But I curios if we can automatically remove unused splat since the splat in this implementation grows more than nerfstudio's splat. 
> 
> I hope it helps make it more efficient in PLY size.

> I hope it helps make it more efficient in PLY size.

@ichsan2895 I had tried to detect unused guassians during training and found that they only contribute to a very small portion of total number of gaussians. To reduce the PLY size, there are a few ideas:
(1) one is to set a higher culling threshold.
(2) set max SH levels from 3 to 2.

File: nerfstudio/models/gaussian_splatting.py (hunk 7 above)
Comment (jb-ye) [reply nerfstudio/models/gaussian_splatting.py:694]: My implementation was trying to mimic an implementation that is closer enough to the original GSplat paper. I can potentially add it back as an option. Do you have examples that show its effectiveness in terms of removing huge spikey gaussians?

**Comment by ichsan2895** [comment]
> > But I curios if we can automatically remove unused splat since the splat in this implementation grows more than nerfstudio's splat.
> > I hope it helps make it more efficient in PLY size.
> 
> > I hope it helps make it more efficient in PLY size.
> 
> @ichsan2895 I had tried to detect unused guassians during training and found that they only contribute to a very small portion of total number of gaussians. To reduce the PLY size, there are a few ideas: (1) one is to set a higher culling threshold. (2) set max SH levels from 3 to 2.

So far, cull_alpha_thresh: float = 0.1 is more efficient than 0.005. Please see my experiment in code review

**Comment by ichsan2895** [comment]
> > But I curios if we can automatically remove unused splat since the splat in this implementation grows more than nerfstudio's splat.
> > I hope it helps make it more efficient in PLY size.
> 
> > I hope it helps make it more efficient in PLY size.
> 
> @ichsan2895 I had tried to detect unused guassians during training and found that they only contribute to a very small portion of total number of gaussians. To reduce the PLY size, there are a few ideas: (1) one is to set a higher culling threshold. (2) set max SH levels from 3 to 2.

So far, cull_alpha_thresh: float = 0.1 is more efficient than 0.005. Please see my experiment in code review

File: nerfstudio/models/gaussian_splatting.py (hunk 7 above)
Comment (kerrj) [reply nerfstudio/models/gaussian_splatting.py:694]: I think leaving it off by default with an option makes sense, the effect is mostly visible in less well reconstructed parts of the scene, for example the edges of this rose plant: 
Without regularization:
Screenshot 2024-01-06 at 3 28 35 PM
With regularization:
Screenshot 2024-01-06 at 3 28 53 PM
Less elongated gaussians also makes downstream usage easier for cropping the scene and reduces popping artifacts slightly, here's some videos:
Without:

https://github.com/nerfstudio-project/nerfstudio/assets/15806475/39682096-e778-4693-af6d-398c88bdb42e

With:

https://github.com/nerfstudio-project/nerfstudio/assets/15806475/8331c341-d14c-4e07-9b11-d51b48713214

**Comment by jb-ye** [comment]
@kerrj My update:
1. The "PSNR drops every 5k steps while training" you observed is a bug in evaluation code. This line: https://github.com/nerfstudio-project/nerfstudio/blob/main/nerfstudio/data/datamanagers/full_images_datamanager.py#L342 should be 
```
        image_indices = [i for i in range(len(self.eval_dataset))]
```
because `self.eval_unseen_cameras` was updated by next_eval() every 100 iterations. I will fix this bug as part of this CR.

2. I experimented with `cull_alpha_thresh=0.1`, PSNR drops about average ~0.3dB on a large benchmark. Looks good to me. I will set it to 0.1 by default. One thing that could be a bit of concerns is with a high threshold, the first gaussian culling operation will eliminate a significant portion of initialized gaussians (up to 80% on some datasets). I am not sure if this is desired. 

3. I will add back the SH regularization as implemented earlier as an option, and will be off by default. 

Please let me know if you have other comments in mind. I plan to update this PR early next week according to the above three items.

**Comment by ichsan2895** [comment]
> @kerrj My update:
> 
>     1. The "PSNR drops every 5k steps while training" you observed is a bug in evaluation code. This line: https://github.com/nerfstudio-project/nerfstudio/blob/main/nerfstudio/data/datamanagers/full_images_datamanager.py#L342 should be
> 
> 
> ```
>         image_indices = [i for i in range(len(self.eval_dataset))]
> ```
> 
> because `self.eval_unseen_cameras` was updated by next_eval() every 100 iterations. I will fix this bug as part of this PR.
> 
>     2. I experimented with `cull_alpha_thresh=0.1`, PSNR drops about average ~0.5dB on a large benchmark. Looks good to me. I will set it to 0.1 by default. One thing that could be a bit of concerns is with a high threshold, the first gaussian culling operation will eliminate a significant portion of initialized gaussians (up to 80% on some datasets). I am not sure if this is desired.
> 
>     3. I will add back the SH regularization as implemented earlier as an option, and will be off by default.
> 
> 
> Please let me know if you have other comments in mind. I plan to update this PR early next week according to the above three items.

I have some test in desolation, playroom, and truck dataset, I have no problem with elimination of significant portion of initialized gaussians. In the end at 30k steps, it will grows too about 50-75% original gaussian count (`culling alpha threshold=0.005`).

image
image
image
image

**Comment by jb-ye** [comment]
@ichsan2895 because of a bug, the metrics you see in the plot is not accurate unfortunately. You will need to use `ns-eval` to get the correct number of numbers rather than grep the number from wandb/tensorflow plot.

**Review by kerrj — CHANGES_REQUESTED** [review]
Looks good thank you! Just one last thing; can we tweak `refinement_after` to continue culling past 15k steps? This can reduce the number of gaussians quite a bit (on bicycle, from 3.1M to 2.4M) and doesn't affect quality noticably in my experience.

**Comment by jb-ye** [comment]
@kerrj Let me experiment a bit more on your suggestions.

**Review by kerrj — APPROVED** [review]

**Comment by kerrj** [comment]
Is there anything else you wanted to add or is it ok if I go ahead and merge it?

**Comment by ichsan2895** [comment]
> Is there anything else you wanted to add or is it ok if I go ahead and merge it?

LGTM so far 👍

**Pull request merged**